src/traefik.py

```diff
@@ -0,0 +1,574 @@
+                "service": service_name,
+                "entryPoints": ["websecure"],
+                "tls": tls_entry,
+            }
+        }
+
+    def pull_tcp_entrypoints(self) -> Dict[str, int]:
+        """Pull the currently configured tcp entrypoints from the static config."""
+        try:
+            static_config_raw = self._container.pull(STATIC_CONFIG_PATH).read()
+        except PathError as e:
+            logger.error(f"Could not fetch static config from container; {e}")
+            return {}
+
+        static_config = yaml.safe_load(static_config_raw)
+        eps = static_config["entryPoints"]
+        return {k: v for k, v in eps.items() if k not in {"diagnostics", "web", "websecure"}}
+
+    @property
+    def is_running(self):
+        """Whether the traefik service is running."""
+        if not self._container.can_connect():
+            return False
+        return bool(self._container.get_services(self._service_name))
+
+    def restart(self):
+        """Restart the pebble service."""
+        layer = {
+            "summary": "Traefik layer",
+            "description": "Pebble config layer for Traefik",
+            "services": {
+                self._service_name: {
+                    "override": "replace",
+                    "summary": "Traefik",
+                    # trick to drop the logs to a file but also keep them available in the pod logs
+                    "command": '/bin/sh -c "{} | tee {}"'.format(BIN_PATH, self.log_path),
+                    "startup": "enabled",
+                },
+            },
+        }
+
+        if not self.is_running:
+            self._container.add_layer(self._layer_name, cast(LayerDict, layer), combine=True)
+            logger.debug(f"replanning {self._service_name!r} after a service update")
+            self._container.replan()
+        else:
+            logger.debug(f"restarting {self._service_name!r}")
+            self._container.restart(self._service_name)
+
+    def clear_dynamic_configs(self):
+        """Delete all yamls from the dynamic config dir."""
+        self._container.exec(["find", DYNAMIC_CONFIG_DIR, "-name", "*.yaml", "-delete"])
+        logger.debug("Deleted all dynamic configuration files.")
+
+    def delete_dynamic_config(self, file_name: str):
+        """Delete a specific yaml from the dynamic config dir."""
+        self._container.remove_path(Path(DYNAMIC_CONFIG_DIR) / file_name)
+        logger.debug("deleted dynamic configuration file: %s", file_name)
+
+    def add_dynamic_config(self, file_name: str, config: str):
+        """Push a yaml to the dynamic config dir."""
+        self._container.push(Path(DYNAMIC_CONFIG_DIR) / file_name, config, make_dirs=True)
+
+        logger.debug("Updated dynamic configuration file: %s", file_name)
+
+    def clear_certs(self):
+        """Delete the server cert and server key."""
+        self._container.remove_path(SERVER_CERT_PATH)
+        self._container.remove_path(SERVER_KEY_PATH)
```

Comment:
We should be mindful here of potential race conditions.
Traefik watches these and reloads automatically.
We should make sure the caller first updates traefik config, and only then deletes the certs.
Should document this as well.

Reply:
I was planning to address more of this in a next PR adding filesystem state manifest-based "as atomic as possible" transactions.
I see the danger but I don't think that's in scope for this PR, unless you think I introduced a bug with my changes? I tried to keep the logic as unchanged as possible, this pr should, if I did things right, be a refactoring only.

Reply:
that said I'm fine adding a todo or comments to point this out to our tomorrow selves

Reply:
beside, I just realized that in `main`, the charm method calling what is now this code [is unreachable](https://github.com/canonical/traefik-k8s-operator/blob/main/src/charm.py#L322)

so the `clear_certs` code is effectively unused as it currently stands, and I'll remove it from here (and the two unreachable handlers in charm.py).

The concern you raised still applies, but to the `remove_ca` method. I'll add a note there.